tensorly/decomposition/_nn_cp.py

```diff
@@ -294,17 +333,36 @@ def non_negative_parafac_hals(
     # Avoiding errors
     for fixed_value in fixed_modes:
         sparsity_coefficients[fixed_value] = None
+        ridge_coefficients[fixed_value] = None
 
     for mode in range(n_modes):
-        if sparsity_coefficients[mode] is not None:
+        if (sparsity_coefficients[mode] is not None) and (mode not in nn_modes):
             warnings.warn("Sparsity coefficient is ignored in unconstrained modes.")
+
     # Generating the mode update sequence
     modes = [mode for mode in range(n_modes) if mode not in fixed_modes]
 
-    # initialisation - declare local varaibles
+    # initialisation - declare local variables
     rec_errors = []
 
-    # Iteratation
+    # Changing None values to 0 in regularization coefficients
+    for i in range(n_modes):
+        if ridge_coefficients[i] is None:
+            ridge_coefficients[i] = 0
+        if sparsity_coefficients[i] is None:
+            sparsity_coefficients[i] = 0
+
+    if callback is not None:
+        cp_tensor = CPTensor((weights, factors))
+        loss = (tl.norm(tensor - tl.cp_to_tensor(cp_tensor)) ** 2) / 2
+        rec_error = tl.sqrt(2 * loss) / norm_tensor
+        for mode, factor in enumerate(factors):
+            loss += ridge_coefficients[mode] ** 2 * tl.norm(
+                factor
+            ) ** 2 + sparsity_coefficients[mode] * tl.sum(tl.abs(factor))
+        callback(cp_tensor, rec_error, loss)
```

Comment:
This is very specific computation just to pass to a callback no? What are the actual cases? I'd almost just have a conditional here if needed instead.